### PR TITLE
Onboarding wizard: fix persistence of previously selected template mode when user revisits wizard

### LINKED ISSUE #5024: Onboarding wizard: Next button sometimes incorrectly disabled
## Bug Description

As brought up on Slack:

Screen Shot 2020-07-10 at 12 07 58 PM

Bug is probably somewhere in here: 

https://github.com/ampproject/amp-wp/blob/develop/assets/src/onboarding-wizard/pages/template-mode/index.js#L32-L46

## Expected Behaviour

If the user has completed the wizard previously, their mode should be preselected and they should be able to move forward immediately.

## Steps to reproduce

1. Complete the onboarding wizard
2. Revisit the onboarding wizard
3. Change your answer to the technical question and move forward
4. Next button is incorrectly disabled

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Summary

Fixes #5024 

Earlier in the onboarding wizard project, it was decided that when a user who previously configured the plugin revisits the wizard, the following scenarios could occur:

1) If they change their answer to the technical question, the template mode would be unselected and we'd show text such as "Previously selected" on the previously set template mode.
2) If they don't change their answer, we persist the previous template mode selection.

Somewhere in last week's large PR for the settings screen there was a regression in this behavior, reported in #5024. This fixes it, adding an E2E test covering the scenarios above, plus a third scenario: If a revisiting user keeps the technical question the same, changes their template mode, and goes back and changes their answer to the technical question, we don't override anything and instead persist the choice they had just made.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: assets/src/onboarding-wizard/components/template-mode-override-context-provider.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	const { developerToolsOption, fetchingUser, originalDeveloperToolsOption } = useContext( User );
+	const [ respondedToDeveloperToolsOptionChange, setRespondedToDeveloperToolsOptionChange ] = useState( false );
+
+	const { theme_support: themeSupport } = editedOptions || {};
+	const { theme_support: originalThemeSupport } = originalOptions || {};
+
+	const [ readerModeWasOverridden, setReaderModeWasOverridden ] = useState( false );
+
+	/**
+	 * Override with transitional if the user has selected reader mode and their currently active theme as reader theme.
```
Comment (pierlon) [inline assets/src/onboarding-wizard/components/template-mode-override-context-provider.js]:
```suggestion
	 * Override with transitional if the user has selected reader mode and their currently active theme is the same as the selected reader theme.
```

File: assets/src/onboarding-wizard/components/template-mode-override-context-provider.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	useEffect( () => {
+		if ( fetchingUser ) {
+			return;
+		}
+
+		if ( respondedToDeveloperToolsOptionChange ) {
+			return;
+		}
+
+		// If user has already made a change, don't do anythhing.
```
Comment (pierlon) [inline assets/src/onboarding-wizard/components/template-mode-override-context-provider.js]:
```suggestion
		// If user has already made a change, don't do anything.
```

File: assets/src/onboarding-wizard/components/template-mode-override-context-provider.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		fetchingUser,
+		originalDeveloperToolsOption,
+		originalThemeSupport,
+		respondedToDeveloperToolsOptionChange,
+		themeSupport,
+		updateOptions,
+	] );
+
+	return (
+		<ReaderModeOverride.Provider value={ readerModeWasOverridden }>
```
Comment (pierlon) [inline assets/src/onboarding-wizard/components/template-mode-override-context-provider.js]: Shouldn't this be `TemplateModeOverride`? Any advantages of having `ReaderModeOverride` only handling `readerModeWasOverridden`?

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/onboarding-wizard/components/template-mode-override-context-provider.js (hunk 3 above)
Comment (johnwatkins0) [reply assets/src/onboarding-wizard/components/template-mode-override-context-provider.js]: Yup, you're right, I missed that one.

**Review by pierlon — APPROVED** [review]

**Comment by westonruter** [comment]
I'm seeing something unexpected: 

0. Open wizard and leave technical question selection unchanged. Click next.

> image

1. See the template mode pre-selected as expected:

> image

2. Click previous. Change the technical question answer:

> image

3. Click next and see “Previously selected” as expected with no template mode selected, as expected.

> image

4. Click previous _again_ and toggle the technical question back:

> image

5. Now there is no template mode selected _and_ no mode that has “Previously selected”:

> image

Expected that a mode should either me pre-selected _or_ to have the “Previously selected” label next to it.

**Review by westonruter — CHANGES_REQUESTED** [review]
I'm seeing something unexpected: 

0. Open wizard and leave technical question selection unchanged. Click next.

> image

1. See the template mode pre-selected as expected:

> image

2. Click previous. Change the technical question answer:

> image

3. Click next and see “Previously selected” as expected with no template mode selected, as expected.

> image

4. Click previous _again_ and toggle the technical question back:

> image

5. Now there is no template mode selected _and_ no mode that has “Previously selected”:

> image

Expected that a mode should either be pre-selected _or_ to have the “Previously selected” label next to it.

**Review by westonruter — APPROVED** [review]

**Pull request merged**